aip/0166.md

```diff
@@ -0,0 +1,86 @@
+---
+aip:
+  id: 166
+  state: reviewing
+  created: 2020-06-24
+permalink: /166
+redirect_from:
+  - /0166
+---
+
+# Hierarchically-derived resources
+
+Services occasionally have resources where the resource's effective values are
+determined by consulting both values on the resource as well as values
+elsewhere, such as a parent in the hierarchy, and "flattening" them together.
+
+This is most common for policies, which could be applied against a resource, a
+collection of resources, or a "container" resource (such as a project or
+organization).
+
+In these situations, users need to distinguish between properties applied to
+the resource directly and properties that are inherited.
+
+## Guidance
+
+Services **should** return only field values on the resource when responding to
+`Get` and `List` methods by default.
+
+Additionally, services **may** add a field to `Get` and `List` requests that
+allows the user to opt-in to receiving the effective configuration instead:
+
+```proto
+message GetBookRequest {
+  // The name of the book.
+  string name = 1 [(google.api.resource_reference) = {
+    type: "library.googleapis.com/Book"
+  }];
+
+  // If true, values that are set on other objects in the hierachy and
+  // inherited by the book will be included.
+  // If false (default), only fields directly set on the book are included.
+  bool effective = 2;
+}
+```
+
+Services **must not** add this field for cases where determining the effective
+values is expensive.
+
+**Note:** Services **may** make it cheap to load effective values by utilizing
+eventual consistency for inherited values. However, services **should** ensure
+that values set directly on the resource being queried are strongly consistent.
+
+### Expensive computations
+
+On rare occasion, the effective values for a resource can be expensive to
+compute and not referenced often. Additionally, a service might need to charge
+for the query.
+
+In these situations, the service **may** provide a specialized method:
```

Comment:
I think this should apply where the cost of the effective computation is substantially different than the cost of service the non-effective resource (not just where it is more expensive).

For example; if a separate database were used to store a precomputed effective policy efficiently for policy evaluation, that would be much cheaper to perform the lookup against (by design), and I would want to reference it from a separate API method so I could easily maintain separate SLO/quota/consistency-guarantees for each method (both in implementation and user-obvious documentation).